packages/plugin-ext/src/plugin/lm-ext.ts

```diff
@@ -0,0 +1,159 @@
+    isMcpHttpServerDefinitionDto,
+} from '../common/lm-protocol';
+import { PLUGIN_RPC_CONTEXT } from '../common/plugin-api-rpc';
+import { PluginPackageMcpServerDefinitionProviderContribution } from '../common';
+import { PluginLogger } from './logger';
+import { McpHttpServerDefinition, McpServerDefinition, URI } from './types-impl';
+
+// Local interfaces that match the proposed MCP API
+interface McpServerDefinitionProvider {
+    readonly onDidChangeMcpServerDefinitions?: theia.Event<void>;
+    provideMcpServerDefinitions(): theia.ProviderResult<McpServerDefinition[]>;
+    resolveMcpServerDefinition?(server: McpServerDefinition): theia.ProviderResult<McpServerDefinition>;
+}
+
+export class LmExtImpl implements McpServerDefinitionRegistryExt {
+
+    private proxy: McpServerDefinitionRegistryMain;
+    private logger: PluginLogger;
+    private readonly providers = new Map<number, McpServerDefinitionProvider>();
+    private readonly providerChangeListeners = new Map<number, Disposable>();
+    private handleCounter = 0;
+    private announcedMCPProviders: string[] = [];
+
+    constructor(protected readonly rpc: RPCProtocol) {
+        this.proxy = this.rpc.getProxy(PLUGIN_RPC_CONTEXT.MCP_SERVER_DEFINITION_REGISTRY_MAIN);
+        this.logger = new PluginLogger(this.rpc, 'lm');
+    }
+
+    registerMcpServerDefinitionProvider(id: string, provider: McpServerDefinitionProvider): theia.Disposable {
+        if (this.announcedMCPProviders.indexOf(id) === -1) {
+            this.logger.warn(`An unknown McpProvider tried to register, please check the package.json: ${id}`);
+        }
+        const handle = this.handleCounter++;
+        this.providers.set(handle, provider);
+
+        this.proxy.$registerMcpServerDefinitionProvider(handle, id);
+
+        // Listen for changes from the provider
+        if (provider.onDidChangeMcpServerDefinitions) {
+            const changeListener = provider.onDidChangeMcpServerDefinitions(() => {
+                this.proxy.$onDidChangeMcpServerDefinitions(handle);
+            });
+            this.providerChangeListeners.set(handle, changeListener);
+        }
+
+        return Disposable.create(() => {
+            this.providers.delete(handle);
+            const changeListener = this.providerChangeListeners.get(handle);
+            if (changeListener) {
+                changeListener.dispose();
+                this.providerChangeListeners.delete(handle);
+            }
+            this.proxy.$unregisterMcpServerDefinitionProvider(handle);
+        });
+    }
+
+    async $provideServerDefinitions(handle: number): Promise<McpServerDefinitionDto[]> {
+        const provider = this.providers.get(handle);
+        if (!provider) {
+            return [];
+        }
+
+        try {
+            const definitions = await provider.provideMcpServerDefinitions();
+            if (!definitions) {
+                return [];
+            }
+
+            return definitions.map(this.convertToDto);
```

Comment:
```suggestion
            return definitions.map(def => this.convertToDto(def));
```

Reply:
Might be dangerous, as a future change in `convertToDto` may expect `this` to work.